examples/cpp/dds/CustomPayloadPoolExample/CustomPayloadPoolPublisher.cpp

```diff
@@ -0,0 +1,211 @@
+// Copyright 2023 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file CustomPayloadPoolPublisher.cpp
+ *
+ */
+
+#include "CustomPayloadPoolPublisher.h"
+#include <fastrtps/attributes/ParticipantAttributes.h>
+#include <fastrtps/attributes/PublisherAttributes.h>
+#include <fastdds/dds/domain/DomainParticipantFactory.hpp>
+#include <fastdds/dds/publisher/Publisher.hpp>
+#include <fastdds/dds/publisher/qos/PublisherQos.hpp>
+#include <fastdds/dds/publisher/DataWriter.hpp>
+#include <fastdds/dds/publisher/qos/DataWriterQos.hpp>
+#include <rtps/history/TopicPayloadPoolRegistry.hpp>
+
+#include <thread>
+
+using namespace eprosima::fastdds::dds;
+
+CustomPayloadPoolPublisher::CustomPayloadPoolPublisher()
+    : participant_(nullptr)
+    , publisher_(nullptr)
+    , topic_(nullptr)
+    , writer_(nullptr)
+    , type_(new CustomPayloadPoolPubSubType())
+{
+    matched_ = 0;
+    firstConnected_ = false;
+}
+
+bool CustomPayloadPoolPublisher::init()
+{
+    hello_.index(0);
+    hello_.message("CustomPayloadPool");
+    DomainParticipantQos pqos = PARTICIPANT_QOS_DEFAULT;
+    pqos.name("Participant_pub");
+    auto factory = DomainParticipantFactory::get_instance();
+
+    participant_ = factory->create_participant(0, pqos);
+
+    if (participant_ == nullptr)
+    {
+        return false;
+    }
+
+    //REGISTER THE TYPE
+    type_.register_type(participant_);
+
+    //CREATE THE PUBLISHER
+    PublisherQos pubqos = PUBLISHER_QOS_DEFAULT;
+
+    publisher_ = participant_->create_publisher(
+        pubqos,
+        nullptr);
+
+    if (publisher_ == nullptr)
+    {
+        return false;
+    }
+
+    //CREATE THE TOPIC
+    TopicQos tqos = TOPIC_QOS_DEFAULT;
+
+    topic_ = participant_->create_topic(
+        "CustomPayloadPoolTopic",
+        "CustomPayloadPool",
+        tqos);
+
+    if (topic_ == nullptr)
+    {
+        return false;
+    }
+
+    // CREATE THE WRITER
+    DataWriterQos wqos = DATAWRITER_QOS_DEFAULT;
+
+    // CREATE CUSTOM TOPIC PAYLOAD POOL
+    payload_pool = std::make_shared<PayloadPool>();
```

Comment:
Isn't the intention of the example to show how to use the same pool instance for both writer and reader? In that case, this should be instantiated in the main and injected to both publisher and subscriber